Complete documentation rewrite

Luc45 opened the pull request:
**PS: Greg, please, don't review this on Saturday, nor Sunday.**

Definetly not done by O1 Pro. Trust me bro.

### What Changed
- Complete reorganization of the documentation structure
- New sidebar navigation
- Consolidated duplicate content (CLI/Dashboard stuff was everywhere)
- Added missing sections for advanced usage and test execution
- Made everything more consistent and interconnected

### Testing Instructions
- Run this branch locally
- See the difference in the organization in the sidebar, see if you like it
- Navigate the documentation
- Compare this proposed documentation with the old documentation, preferably PAGE BY PAGE, as there are some pages in the old documentation with graphics and things like that that were not ported over.. Ideally, we should try to make this new documentation a bit more visual in some parts, with some graphs, etc

### Known TODOs
- Port over graphics from old docs
- Add more visuals where it makes sense (especially for complex concepts)
- Double-check all cross-links work
- Make sure we didn't lose any crucial info in the reorganization

O1 is good but I had to stick with it for over 5h until 2am because I didn't want to lose the huge context I had while generating this doc. I generated this by first figuring out a structure that made sense with it, then generating it page by page, then today, in the morning, coming up with all kinds of crazy scripts to fix the internal links. O1 is good, but I just don't want to make it feel like it's magic. There was a looot of work.

Luc45 committed:
Add comprehensive documentation for QIT CLI

- Introduced a new combined documentation file covering the QIT testing platform.
- Detailed installation instructions and requirements for using the QIT CLI.
- Provided a quick start guide with essential commands to run tests.
- Explained various test types available and how to execute them.
- Included information on integrating QIT with GitHub workflows for automated testing.
- Added sections on custom E2E tests, tagging, and running tests against plugins/themes.
- Documented lifecycle phases of custom tests and environment setup procedures.

Luc45 committed:
Remove outdated custom tests docs

Deleted several markdown files related to custom E2E tests, including introduction, generating tests, tagging tests, and security architecture. This cleanup helps streamline documentation and remove early access content that is no longer relevant.

Luc45 committed:
Add advanced config handlers and scripting docs

- Introduced new documentation for advanced configuration with custom handlers, detailing their use cases and examples.
- Added a section on scripting to automate QIT CLI tasks, including sample scripts for Bash, NPM, and Composer.
- Removed outdated getting started guide and several related environment setup documents to streamline content.

Luc45 committed:
Remove outdated CLI documentation

Deleted all documentation related to the CLI, including installation instructions, running tests, useful commands, scripting examples, GitHub workflows, and notification settings. This cleanup helps streamline the project by removing obsolete content.

Luc45 committed:
Update contact info formatting in support docs

Changed email addresses to use backticks for better readability. Clarified instructions for reporting security vulnerabilities.

Luc45 committed:
Add orchestration documentation

- Created a new page on orchestration for E2E testing.
- Explained shared vs. isolated setups and their importance.
- Detailed the process of managing multiple plugins in tests.
- Updated lifecycle docs to link to the new orchestration page.
- Added orchestration to the sidebar for easy navigation.

Luc45 committed:
Remove combined documentation files

Deleted the combined documentation file that contained extensive information about the QIT testing platform, including installation guides, usage instructions, and custom test features. This cleanup helps streamline the documentation structure for better maintainability.

Luc45 committed:
Add PHPCompatibility

Luc45 committed:
Add PHP compatibility to managed tests

- Included 'managed-tests/phpcompatibility' in the sidebar
- Enhances testing coverage for PHP compatibility issues

Luc45 committed:
Update links and improve documentation clarity

- Updated internal links to point to the correct files.
- Removed redundant text and ensured consistency across documents.
- Enhanced descriptions for better understanding of features and next steps.

Luc45 committed:
Update running tests documentation

Clarified the process for running custom E2E tests.
- Emphasized local execution for rapid iteration.
- Explained automatic environment management during test runs.
- Added details on SUT, additional plugins, and dependencies.
- Highlighted cloud testing triggered by marketplace submissions.

Luc45 committed:
Refine installation guide for QIT CLI

Updated the Quick Start Guide in the intro documentation to recommend global installation of QIT CLI via Composer, ensuring easier access and consistency across projects. Detailed steps for adding Composer's `bin` directory to the `PATH` and simplified command usage are now included. Retained instructions for per-project installation as an alternative, but emphasized the benefits of the global approach.

Luc45 committed:
Enhanced docs for multi-plugin E2E testing

- Added detailed guidance on setting up compatibility tests in existing documentation.
- Created a new document explaining the nuances of `test`, `bootstrap`, and `activate` actions during plugin compatibility testing.
- Updated lifecycle management sections across multiple documents to clarify shared and isolated phase interactions during tests.

Luc45 committed:
Typo

Luc45 committed:
Enhanced e2e test command readability

Updated the end-to-end testing commands in the documentation to improve clarity and added alternative CLI parameter usage. This change:
- Switches code block syntax highlighting from `bash` to `qitbash`.
- Introduces a simpler command format for quick execution.
- Provides an expanded version with detailed CLI parameters, enhancing understandability for users preferring explicit configurations.

Luc45 committed:
Remove

Luc45 committed:
Refine compatibility test docs for clarity

- Removed suggestion to create an advanced guide
- Streamlined the section on testing nuances

Luc45 committed:
Enhance docs with compatibility test section

Added a new section on compatibility tests to the custom E2E testing introduction. This update includes guidance on using shared setup and teardown scripts for orchestrating complex scenarios across multiple plugins, aiming to ensure consistent environments and results in testing processes.

Luc45 committed:
Remove unnecessary footers

Luc45 committed:
Enhance doc links and update code blocks

- Updated markdown links for better navigation in documentation
- Changed bash code blocks to qitbash for specific syntax highlighting

singerb commented on docs/managed-tests/plugin-check.md:
```diff
@@ -0,0 +1,35 @@
+# Plugin Check Tests
```
I'll be the AI skeptic and say I don't love this one; it's both longer and more verbose than the original, while not containing all the same info (which specific checks do we run, and the link to the review guidelines).

Luc45 replied:
I've incorporated all the info from the old one, while keeping the "writing style" consistent with the rest of the documentation - which is one of the good things about this rewrite, to consolidate the style of writing. I'm open to any suggestions you might have, and feel free to own this and change it however you see fit, it's completely fine. Let me know if you like this version or not:

![image](https://github.com/user-attachments/assets/b797dd4d-c946-46a3-a50a-0b430c1ffaa5)

singerb commented on docs/managed-tests/plugin-check.md:
```diff
@@ -0,0 +1,35 @@
+# Plugin Check Tests
+
+Plugin Check tests run the [wordpress.org automated plugin check](#) tool against your extension. This tool helps identify issues and guidelines violations aligned with the WordPress.org plugin directory standards. While this test is optional and not automatically triggered by QIT for WooCommerce Marketplace submissions, it can be valuable if you also distribute your extension on wordpress.org or ...
```
Missing link.

Luc45 replied:
Fixed

singerb commented on docs/managed-tests/security.md:
```diff
@@ -0,0 +1,50 @@
+# Security Tests
```
Same comment on this one as the plugin check; I think the original was better.

Luc45 replied:
Agree 100%, that was an oversight. Let me know what you think of this version, and similar to the Plugin Check, feel free to modify it in any way shape or form.

![image](https://github.com/user-attachments/assets/4481fcc2-3bf0-456f-958d-031818c9eed8)

singerb commented on docs/managed-tests/validation.md:
```diff
@@ -0,0 +1,34 @@
+
+- **Success:** All required metadata is present and correct, and no invalid headers or incompatible feature declarations were found.
+- **Warning:** Some headers may be missing or incorrect, or you declared incompatibility with features that you should ideally support.
+- **Failed (for Themes):** Outdated templates or critical metadata issues prevent passing the validation.
```
This is inaccurate, it's a warning right now.

Luc45 replied:
Fixed, check screenshot below

singerb commented on docs/managed-tests/validation.md:
```diff
@@ -0,0 +1,34 @@
+# Validation Tests
```
Again, missing links here. This rewrite is arguably ok, as the original is perhaps a bit too verbose, but I'm still iffy.

Luc45 replied:
Let me know what you think of this one (Updated screenshot just now, 14h30 UTC):
![image](https://github.com/user-attachments/assets/d8df8816-b390-44ca-98ed-d5aa819b659d)

singerb commented:
Oh, one overall note: I thought we wanted things `In sentence case` rather than `In Title Case`, but PCYsg-5n-p2me that title-case is fine. This means the PR is probably fine, but if you can feed the style guide to the AI too maybe it can follow it automatically?

zhongruige commented on docs/core-concepts/test-types-overview.md:
```diff
@@ -0,0 +1,62 @@
+QIT supports multiple kinds of tests designed to ensure the quality, compatibility, and security of WooCommerce extensions. These tests fall into two main categories: **Managed Tests** and **Custom E2E Tests**.
+
+Why does this matter?  
+By familiarizing yourself with these test types, you can leverage QIT to catch issues early, maintain compatibility with evolving WordPress and WooCommerce versions, and deliver a reliable experience to merchants and their customers.
```
I would recommend re-working this a bit, and drop the `Why does this matter?` heading, so it's something like this instead:

```suggestion
By utilizing test types provided by QIT, you can leverage QIT to catch issues early, maintain compatibility with evolving WordPress and WooCommerce versions, and deliver a reliable experience to merchants and their customers.
```

zhongruige commented on docs/core-concepts/test-types-overview.md:
```diff
@@ -0,0 +1,62 @@
+
+## Managed Tests
+
+Definition:  
```
I would drop the "Definition" header here and below, it feels redundant.

zhongruige committed:
Wording tweak

zhongruige committed:
Fix tunneling wording on activation test page

Luc45 committed:
Use qitbash where applicable

Luc45 committed:
Fix all missing links

Luc45 committed:
"In sentence case" for headings

Luc45 committed:
"In sentence case" for bullet items

Luc45 commented:
> Oh, one overall note: I thought we wanted things `In sentence case` rather than `In Title Case`, but PCYsg-5n-p2me that title-case is fine. This means the PR is probably fine, but if you can feed the style guide to the AI too maybe it can follow it automatically?

Fixed in https://github.com/woocommerce/qit-documentation/pull/50/commits/802a27ef5d976ad88cea0650460ba2b73bc241fe and https://github.com/woocommerce/qit-documentation/pull/50/commits/307844f61c3878b13a51134c17332fae01e40217

Luc45 committed:
Rewrite plugin check tests

Luc45 committed:
Update security test docs

Luc45 committed:
Validation test write-up

Luc45 committed:
Update link

Luc45 committed:
Tweaks

Luc45 committed:
PHPCompatibility page

Luc45 committed:
Remove "next steps" from managed tests docs

Luc45 committed:
Removed all "Next steps" at the bottom

Luc45 commented:
I've also removed all "Next Steps" sections at the bottom of all pages. These links were disrupting the flow of the documentation and often linking to places that didn't make much sense. Now we let "Next" and "Previous" component at the bottom be the "guide":

![image](https://github.com/user-attachments/assets/dd19dd47-3f88-4d12-a0fd-614e513a1c11)

I've only left a "Next Steps" on the "Installing" page, where it actually makes sense, since the next steps it points to authenticating and running the first managed test.

One thing that I still want to do (but doesn't need to a blocker to merging) is adding CLI options examples to some of the commands, especially around the managed tests, for instance, on the PHPCompatibility tests, we could mention `--min_php_version` and `--max_php_version`, etc.

Luc45 commented:
> but if you can feed the style guide to the AI too maybe it can follow it automatically?

`<irony>`Hey! That's not how AI works! Do you think things are easy like that...?!`</irony>`

Sometimes we just need to get a little creative, here's the series of prompts:

```
Write a PHP script that will recursively iterate over all files ending in ".md" in the current directory, and replace all headings to use "In sentence case" instead of "In Sentence Case".
 
Example:
 
# Security Tests
 
Becomes # Security Tests
 
## What Tools Are Used? becomes ## What tools are used?
 
etc
```

It gave a script like this:

```php
<?php

/**
 * Convert a heading string to sentence case.
 * Example: "What Tools Are Used?" -> "What tools are used?"
 *          "Security Tests"       -> "Security tests"
 */
function toSentenceCase($string) {
    $string = trim($string);
    // Convert entire string to lowercase, then uppercase first character
    $string = strtolower($string);
    $string = ucfirst($string);
    return $string;
}

/**
 * Process a single markdown file, converting headings to sentence case.
 */
function processFile($filename) {
    $content = file_get_contents($filename);
    $lines = explode("\n", $content);

    foreach ($lines as &$line) {
        // Match headings: One or more '#' followed by whitespace, then heading text
        if (preg_match('/^(#+\s+)(.*)$/', $line, $matches)) {
            $prefix = $matches[1]; // The '#' characters plus the trailing space
            $headingText = $matches[2];

            // Convert the heading text to sentence case
            $headingText = toSentenceCase($headingText);

            // Reassemble the line
            $line = $prefix . $headingText;
        }
    }

    // Save updated content back to the file
    file_put_contents($filename, implode("\n", $lines));
}

// Use recursive directory iteration to find all .md files
$directory = new RecursiveDirectoryIterator('.');
$iterator = new RecursiveIteratorIterator($directory);

foreach ($iterator as $file) {
    if ($file->isFile() && preg_match('/\.md$/i', $file->getFilename())) {
        processFile($file->getPathname());
    }
}

echo "Done processing markdown files.\n";
```

I've run the script and generated a git patch with the diff. Then I fed it the diff of the changes that the script caused, which was around 140kb of text, which is a lot, and asked it to analyse carefully all the changes to see if it's accurate. It gave me this:

```
While converting headings to sentence case, some brand names, acronyms, and terms that should remain capitalized have been inadvertently changed to lowercase. These include:

    Brand and Product Names:
        QIT: This acronym stands for "Quality Insights Toolkit" and is a product name. It should remain consistently capitalized as "QIT," not "Qit."
        WooCommerce: This brand name should remain "WooCommerce," not "woocommerce."
        WordPress: Always capitalized as "WordPress," not "wordpress."
        GitHub: Should remain "GitHub," not "github."
        WooCommerce Marketplace: Should stay "WooCommerce Marketplace," not "woocommerce marketplace."
        Cloudflare: When mentioned, "Cloudflare" should remain properly capitalized, not "cloudflare."
        AWS S3, GCP, or Azure Storage: These are brand/service names and acronyms. "AWS," "GCP," and "Azure Storage" should remain capitalized, not "aws s3," "gcp," or "azure storage."

    Acronyms and Technical Terms:
    These are commonly recognized acronyms and should remain in uppercase:
        CLI (Command Line Interface): Should remain "CLI," not "cli."
        UI (User Interface): Remain "UI," not "ui."
        API (Application Programming Interface): Remain "API," not "api."
        E2E (End-to-End): Commonly written as "E2E," not "e2e."
        PHPCS: Stays "PHPCS," not "phpcs."
        PHPStan: Stays "PHPStan," not "phpstan."
        YAML, JSON: These file format acronyms are usually uppercase. It's common to see "YAML" and "JSON" remain uppercase.
        WP-CLI: This should remain "WP-CLI," not "wp-cli."
        NPM: Typically written as "npm" in lowercase by the community, but if originally uppercase, it might need to remain consistent. The original usage in the documentation was "NPM," which is often stylized as "npm." If the original text used "NPM," then consistency with the original would be ideal, or consider using the official stylization "npm."
        CI/CD: These are standard acronyms. They should remain "CI/CD," not "ci/cd."

    Special Terms and Proper Nouns
        SemGrep: Brand name of a tool, usually capitalized as "Semgrep" or "SemGrep" depending on their official styling. In the original text, it appeared as "SemGrep." It should remain in that exact form.
        PHPCS: As above, always uppercase.
```

Then I had to do some back-and-forth with it to keep acronyms capitalized (anything that was originally two uppercase letters) and brands (WooCommerce, etc). After I (and it) was satisfied with the results I repeated the process for lists (such as `- **Foo**`, thus the two commits linked above.

Luc45 committed:
Tweak validation doc

zhongruige committed:
Tweak on plugin check wording

zhongruige committed:
Tweaks to FAQ page

zhongruige committed:
Minor tweaking to casing on troubleshooting page

Luc45 commented on docs/managed-tests/plugin-check.md:
```diff
@@ -0,0 +1,31 @@
+# wordpress.org Plugin Check tests
+
+Tthe [wordpress.org automated plugin check](https://wordpress.org/plugins/plugin-check) tool against your extension. This optional test focuses on guidelines and patterns important for submitting plugins to the WordPress.org directory.
```
```suggestion
The [wordpress.org automated plugin check](https://wordpress.org/plugins/plugin-check) tool against your extension. This optional test focuses on guidelines and patterns important for submitting plugins to the WordPress.org directory.
```

zhongruige replied:
Ack! Good catch.

Luc45 committed:
Update docs/managed-tests/plugin-check.md

zhongruige commented:
@Luc45 we mention for the custom tests and local env:

> The custom E2E tests feature is available as early access.

I'm thinking we can probably remove this wording now, what do you think?

Luc45 commented:
I think so, too. On it.

Luc45 committed:
Remove early-access notices from docs

Removed all instances of early-access notifications across various documentation pages to streamline content and avoid redundancy. This includes updates in sections about custom E2E tests, environment setup, and tunnel usage.

Luc45 committed:
Merge remote-tracking branch 'origin/24-12/doc-rewrite' into 24-12/doc-rewrite

Luc45 committed:
Updated code blocks to use `qitbash` for clarity

- Changed some instances of ```bash to ```qitbash in documentation
- Ensures consistency and correct highlighting for QIT-specific commands across docs

zhongruige committed:
Tweaks to wccom dashboard section

zhongruige committed:
Tweak to running tests via CLI

zhongruige committed:
Tweak Allure capitalization

Luc45 committed:
Use qitbash consistently

- Wrapped CLI commands in code blocks for better readability.
- Standardized the presentation of command examples across multiple documentation files.
- Ensured consistency in displaying commands, aiding clarity and user experience.

Luc45 committed:
Merge remote-tracking branch 'origin/24-12/doc-rewrite' into 24-12/doc-rewrite

Luc45 committed:
qitbash

Luc45 committed:
qitbash

Luc45 committed:
qitbash

zhongruige reviewed (approved):
Just finished going over this again and I believe it's looking good to merge and deploy. We can continue to tweak as needed, but so far seems the links are working as expected and made and pushed in a few adjustments, so i think we're in a good spot.

singerb committed:
Add a missing bit to PCP; de-bold security test for consistency; reword WPScan test; bold the not in the Woo header bit in validation

singerb reviewed (approved):
Yep, as mentioned yesterday I think the overall structure is much better, and with those tweaks the managed tests docs are looking good again too. I pushed some minor fixes to those pages but otherwise I think it's looking good too.
